src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -388,7 +391,11 @@ private Plan.KeysIteration buildKeysIterationPlan()
         if (orderer != null)
             keysIterationPlan = planFactory.sort(keysIterationPlan, orderer);
 
-        assert keysIterationPlan != planFactory.everything; // This would mean we have no WHERE nor ANN clauses at all
+        // This would mean we have no WHERE nor ANN clauses at all; this can happen in case an index was dropped after the
+        // query was initiated
+        if (keysIterationPlan == planFactory.everything)
+            throw invalidRequest(INDEX_MAY_HAVE_BEEN_DROPPED + StatementRestrictions.REQUIRES_ALLOW_FILTERING_MESSAGE);
```

Comment:
Another option is to fire `IndexNotAvailable`.

Reply:
I think I prefer `InvalidRequestException`. `IndexNotAvailable` sounds to me like it will eventually be available, and it won't. Next best thing would be a dedicated exception, but I'm not sure we need that much.